### PR TITLE
Implement Cursor Hiding While Typing & Cursor Inactive Timeout (#284)

### PR DESCRIPTION
**Pull request by yzy-1**
This pull request addresses the discussion on #284 regarding cursor management in niri. Specifically, I have implemented the second solution: hiding the cursor while typing.

The implementation of cursor timeout (solution 1) would require active polling to check the remaining time, which is not straightforward given our current design where all cursor-related updates are triggered by events. Implementing this approach would involve finding a suitable place in the codebase to handle the periodic checks, which I found challenging due to the existing event-driven architecture.

In contrast, hiding the cursor while typing (solution 2) can be seamlessly integrated as it aligns with the existing event-driven model. This implementation is more straightforward and reduces potential complexity.

Please review the changes and let me know if any further adjustments are needed.

### CODE REVIEW

File: src/input/mod.rs [hunk 1]
```diff
@@ -360,16 +364,15 @@ impl State {
         }
 
         let config = self.niri.config.borrow();
-        let config = &config.input.keyboard;
```
Comment (YaLTeR) [inline src/input/mod.rs:363]: Seems like unrelated changes?

**Review by YaLTeR — COMMENTED** [review]
Seems rather straightforward and simple. As long as it works for the people who want this functionality.

> Implementing this approach would involve finding a suitable place in the codebase to handle the periodic checks, which I found challenging due to the existing event-driven architecture.

Sounds similar enough to the idle notifier code. That one's refreshed on every input event, but the cursor one could be refreshed on every pointer event instead.

**Comment by yzy-1** [comment]
> Sounds similar enough to the idle notifier code. That one's refreshed on every input event, but the cursor one could be refreshed on every pointer event instead.

Thanks for the feedback. You're right that it does have similarities with the idle notifier code. However, I believe cursor hiding should occur after a period of inactivity, which means there would be no corresponding pointer event (or even `input::State::process_input_event` will not be called) during that time.

I tried to implement this here, but the effect is that the cursor wouldn't hide until an input event occurred (like key press or something), essentially making the timeout ineffective.

**Comment by yzy-1** [comment]
I tried it again, checking the cursor timeout in `refresh_pointer_outputs` and now it succeeds and works perfectly fine.

File: niri-config/src/lib.rs [hunk 2]
```diff
@@ -2953,6 +2976,10 @@ mod tests {
             cursor {
                 xcursor-theme "breeze_cursors"
                 xcursor-size 16
+                hide-on-key-press
+                inactive-timeout {
+                    duration-ms 3000
+                }
```
Comment (YaLTeR) [inline niri-config/src/lib.rs:2982]: I think a better way is to replace this block with something like `hide-after-inactive-ms 3000`

File: src/input/mod.rs [hunk 3]
```diff
@@ -307,6 +311,8 @@ impl State {
             }
         }
 
+        self.hide_cursor_if_needed();
```
Comment (YaLTeR) [inline src/input/mod.rs:314]: This should be inside `if pressed {}`

File: src/input/mod.rs [hunk 4]
```diff
@@ -383,6 +392,13 @@ impl State {
         self.niri.bind_repeat_timer = Some(token);
     }
 
+    fn hide_cursor_if_needed(&mut self) {
+        let config = self.niri.config.borrow();
+        if config.cursor.hide_on_key_press {
+            self.niri.pointer_hidden = true;
```
Comment (YaLTeR) [inline src/input/mod.rs:398]:
```suggestion
            self.niri.pointer_hidden = true;
            self.niri.queue_redraw_all();
```

File: src/niri.rs [hunk 5]
```diff
@@ -2663,7 +2665,25 @@ impl Niri {
...
+        let config = &config.cursor.inactive_timeout;
+
+        if config.off {
+            return;
+        }
+
+        let now = Instant::now();
+        let timeout = Duration::from_millis(config.duration_ms as u64);
+
+        if now - self.last_cursor_movement >= timeout {
```
Comment (YaLTeR) [inline src/niri.rs:2679]:
```suggestion
        if self.last_cursor_movement.elapsed() >= timeout {
```

**Review by YaLTeR — CHANGES_REQUESTED** [review]

File: niri-config/src/lib.rs (hunk 2 above)
Comment (yzy-1) [reply niri-config/src/lib.rs:2982]: Should the inactive-timeout block of the previous layer be removed?

What I mean is that if a user needs to disable this feature, they should:

```plain
cursor {
  duration-ms -1 // Oh, I hate Magic Numbers!
}
```

or

```plain
cursor {
  inactive-timeout {
    off
    // hide-after-inactive-ms 3000
  }
}
```

Comment (YaLTeR) [reply niri-config/src/lib.rs:2982]: To disable the timeout, just don't put it in the config. This is consistent with other things like `focus-follows-mouse` and so on. In the struct, it should be an Option, then None would mean it's disabled and Some(u32) means it's enabled with that timeout.

Comment (yzy-1) [reply niri-config/src/lib.rs:2982]: Oh I know what you mean. We can use `Options<T>`. My bad.

**Comment by yzy-1** [comment]
OK. I resolved them all.

**Comment by YaLTeR** [comment]
I did some fixes, and I also had to reimplement the inactivity hiding as a timer after all, since otherwise it would fire late due to `refresh()` simply not being called when nothing is happening. Please test that this still works for you; otherwise it's good to merge.

**Review by YaLTeR — APPROVED** [review]

**Comment by yzy-1** [comment]
It worked perfectly! Thank you!

**Comment by YaLTeR** [comment]
Thanks
